core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -77,25 +80,57 @@
             if (exception[0] == null) {
                 exception[0] = e;
             }
-            rolledBack = true;
-            doRollback(context, transactionManager, transaction);
+
+            try {
+                doRollback(context, transactionManager, transaction);
+            }
+            finally {
+                rolledBack = true;
+            }
         } finally {
             try {
-                if (readOnly) {
-                    doRollback(context, transactionManager, transaction);
+                if (readOnly && !rolledBack) {
+                    try {
+                        doRollback(context, transactionManager, transaction);
+                    }
+                    finally {
+                        rolledBack = true;
+                    }
                 } else if (!rolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
                     doCommit(context, transactionManager, transaction);
                 }
             } catch (TransactionSystemException e) {
                 if (!(e.getOriginalException() instanceof RollbackException)) {
                     Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
-                    doRollback(context, transactionManager, transaction);
+                    if (!rolledBack) {
+                        try {
+                            doRollback(context, transactionManager, transaction);
+                        }
+                        finally {
+                            rolledBack = true;
```

Comment:
This seem not used later.

Reply:
See line 126, we are trying to prevent duplicate rollback.

Aside: Trying for minimal change the previous tested logic. With more time I would refactor into multiple methods to clearly document what each branch of logic is trying to do etc...